lenskit-core/src/main/java/org/lenskit/data/dao/EntityCollectionDAOBuilder.java

```diff
@@ -0,0 +1,155 @@
+    private EntityCollectionBuilder lastBuilder = null;
+    private EntityType last = null;
+
+    /**
+     * Index entities by an attribute.
+     * @param et The entity type.
+     * @param attr The attribute to index.
+     * @return The builder (for chaining).
+     */
+    public EntityCollectionDAOBuilder addIndex(EntityType et, TypedName<?> attr) {
+        Preconditions.checkState(entitySets != null, "build() already called");
+        EntityCollectionBuilder builder = findBuilder(et);
+        builder.addIndex(attr);
+        return this;
+    }
+
+    /**
+     * Get the entity types registered with this builder so far.
+     * @return The entity types registered so far.
+     */
+    public Set<EntityType> getEntityTypes() {
+        return ImmutableSet.copyOf(entitySets.keySet());
+    }
+
+    /**
+     * Add an entity to the DAO.
+     * @param e The entity to add
+     * @return The DAO builder (for chaining).
+     */
+    public EntityCollectionDAOBuilder addEntity(Entity e) {
+        Preconditions.checkState(entitySets != null, "build() already called");
+        EntityType type = e.getType();
+        EntityCollectionBuilder bld = findBuilder(type);
+        bld.add(e);
+
+        return this;
+    }
+
+    private EntityCollectionBuilder findBuilder(EntityType type) {
+        if (type != last) {
+            lastBuilder = entitySets.get(type);
+            last = type;
+            if (lastBuilder == null) {
+                lastBuilder = new EntityCollectionBuilder(type);
+                entitySets.put(type, lastBuilder);
+            }
+        }
+        assert lastBuilder != null;
+        return lastBuilder;
+    }
+
+    /**
+     * Add multiple entities.
+     * @param entities The entity list.
+     * @return The builder (for chaining).
+     */
+    public EntityCollectionDAOBuilder addEntities(Entity... entities) {
+        return addEntities(Arrays.asList(entities));
+    }
+
+    /**
+     * Add multiple entities.
+     * @param entities The entity list.
+     * @return The builder (for chaining).
+     */
+    public EntityCollectionDAOBuilder addEntities(Iterable<Entity> entities) {
+        for (Entity e: entities) {
+            addEntity(e);
+        }
+
+        return this;
+    }
+
+    /**
+     * Derive bare entities from the values in another type of entity.  This method only consults the entities added
+     * so far, so it should be called *after* all other calls to {@link #addEntity(Entity)} and friends.  If an entity
+     * has already been added with the same type and ID as one of the derived entites, it is kept instead of the derived
+     * entity.
+     *
+     * @param derived The derived entity type.
+     * @param source The source entity type.
+     * @param attr The source attribute.
+     * @return The builder (for chaining).
+     */
+    public EntityCollectionDAOBuilder deriveEntities(EntityType derived, EntityType source, TypedName<Long> attr) {
```

Comment:
to make sure I'm clear, this is for derinving items and users from ratings, things like that right?